### ISSUE TITLE
Why can't starting state contain data?

### ISSUE THREAD

**Issue by pleepleus** [korken89/smlang-rs#34]
In the code, it says "because there is no way to propagate it (yet)".  I don't fully understand this.  Could someone more familiar with the reason explain?

**Comment by pleepleus** [comment]
After looking more at the code, I think this is referring to the signature for ::new(), which does not take any arguments so thus there is no way to pass initial data for the state.  I think a simple non-breaking change would be to change this signature to accept initial state data _if_ the initial state holds data.

**Comment by ryan-summers** [comment]
It looks like `codegen.rs` actually generates a function signature [`new_with_state`](https://github.com/korken89/smlang-rs/blob/9f4567b6fb05bd867363bb46385f4c33704fe304/macros/src/codegen.rs#L482), which theoretically allow the starting state to contain data. Have you tried this?

**Comment by dzimmanck** [comment]
@ryan-summers, yes that works, and does provide a work-around if you just create a dummy start state with now data.  But that's ugly.  We can do better with the macro.  We simply detect if the starting state has data, and we tailor the input signature of `sm::new` so that if the starting state has data, it requires the initial data to be provided in the function.

**Comment by dzimmanck** [comment]
I am just flagging this as an enhancement and assigning to myself.  It will be easy to implement.  I plan on working on this Friday.

**Comment by pleepleus** [comment]
Closed by PR #39
